libsigner/src/signer_set.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use blockstack_lib::chainstate::stacks::boot::NakamotoSignerEntry;
+use hashbrown::{HashMap, HashSet};
+use stacks_common::types::chainstate::{StacksAddress, StacksPublicKey};
+use wsts::curve::ecdsa;
+use wsts::curve::point::{Compressed, Point};
+use wsts::state_machine::PublicKeys;
+
+/// A reward set parsed into the structures required by WSTS party members and coordinators.
+#[derive(Debug, Clone)]
+pub struct ParsedSignerEntries {
+    /// The signer addresses mapped to signer id
+    pub signer_ids: HashMap<StacksAddress, u32>,
+    /// The signer ids mapped to public key and key ids mapped to public keys
+    pub public_keys: PublicKeys,
+    /// The signer ids mapped to key ids
+    pub signer_key_ids: HashMap<u32, Vec<u32>>,
+    /// The signer ids mapped to wsts public keys
+    pub signer_public_keys: HashMap<u32, Point>,
+    /// The signer ids mapped to a hash set of key ids
+    /// The wsts coordinator uses a hash set for each signer since it needs to do lots of lookups
+    pub coordinator_key_ids: HashMap<u32, HashSet<u32>>,
+}
+
+/// Parsing errors for `ParsedSignerEntries`
+#[derive(Debug)]
+pub enum Error {
+    /// A member of the signing set has a signing key buffer
+    ///  which does not represent a ecdsa public key.
+    BadSignerPublicKey(String),
+    /// The number of signers was greater than u32::MAX
+    SignerCountOverflow,
+}
+
+impl ParsedSignerEntries {
+    /// Try to parse the reward set defined by `NakamotoSignEntry` into the structures required
+    ///  by WSTS party members and coordinators.
+    pub fn parse(is_mainnet: bool, reward_set: &[NakamotoSignerEntry]) -> Result<Self, Error> {
+        let mut weight_end = 1;
+        let mut signer_key_ids = HashMap::with_capacity(reward_set.len());
+        let mut signer_public_keys = HashMap::with_capacity(reward_set.len());
+        let mut coordinator_key_ids = HashMap::with_capacity(4000);
+        let mut signer_ids = HashMap::with_capacity(reward_set.len());
+        let mut wsts_signers = HashMap::new();
+        let mut wsts_key_ids = HashMap::new();
+        for (i, entry) in reward_set.iter().enumerate() {
+            let signer_id = u32::try_from(i).map_err(|_| Error::SignerCountOverflow)?;
```

Comment:
Shouldn't we just error out of `reward_set.len()` is bigger than u32::MAX? Or even something more realistic, like 4000?

Reply:
This does error out in exactly that case. We could error earlier by doing something like:

```
let _total_len = u32::try_from(reward_set.len()).map_err(|_| Error::SignerCountOverflow)?;
for (i, entry) in reward_set.iter().enumerate() {
    let signer_id = u32::try_from(i).unwrap(); // safe because we checked above
```

But I'd prefer avoiding the `unwrap()` even if we know its safe.